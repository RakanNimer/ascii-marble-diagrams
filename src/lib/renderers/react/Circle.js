import React from 'react'
const Circle = ({ r, children }) => {
  return [
    <circle
      key={`circle`}
      r={r}
      fill="#DDDDDDEE"
      cx={r}
      cy={r}
      textAnchor="middle"
    />,
    <text
      key={`text`}
      x="50%"
      y="60%"
      textAnchor="middle"
      fill="black"
      fontSize="80%"
    >
      {children}
    </text>
  ]
}
export { Circle }
