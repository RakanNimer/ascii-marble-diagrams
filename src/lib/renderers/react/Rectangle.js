import React from 'react'
const Rectangle = ({ width, children }) => {
  const mult = 2
  return (
    <g>
      <rect
        key={`Rectangle`}
        width={width * mult}
        height={width}
        fill="#DDDDDD"
        x={0}
        y={0}
        textAnchor="middle"
      />
      <text key={`text`} x="0" y="60%" fill="black" fontSize="80%">
        {children}
      </text>
    </g>
  )
}
export { Rectangle }
