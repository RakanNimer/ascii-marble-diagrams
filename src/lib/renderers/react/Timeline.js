import React from 'react'

const TimelineSVG = props => {
  const { circleWidth, width } = props
  const arrowWidth = circleWidth / 4
  return (
    <svg width={width} height={100} {...props}>
      <line
        x1={0}
        y1={circleWidth * 2}
        x2={width - arrowWidth * 2}
        y2={circleWidth * 2}
        stroke="black"
        strokeWidth="0.2"
        markerEnd="url(#head)"
      />
      <polygon
        id="Triangle"
        stroke="#979797"
        fill="black"
        transform={`translate(${width - arrowWidth * 2} ${circleWidth * 1.5 +
          arrowWidth})`}
        points={`0 0 ${arrowWidth * 2} ${arrowWidth} 0 ${arrowWidth * 2}`}
      />
    </svg>
  )
}
const Timeline = props => {
  const { circleWidth, width, arrowColor, marginTop } = props
  const arrowWidth = circleWidth
  return (
    <div
      style={{
        zIndex: 0,
        marginTop: marginTop, // circleWidth * 2 - arrowWidth / 4,
        position: 'absolute'
      }}
    >
      <div
        key="line"
        style={{
          position: 'absolute',
          width: width - arrowWidth,
          backgroundColor: 'black',
          height: 2,
          // marginTop: arrowWidth / 4,
          zIndex: 0
          // marginTop: arrowWidth / 2
        }}
      />
      <div
        key="arrow"
        style={{
          position: 'absolute',
          width: 0,
          height: 0,
          borderTop: `${arrowWidth / 4}px solid transparent`,
          borderBottom: `${arrowWidth / 4}px solid transparent`,
          borderRight: `${arrowWidth / 4}px solid transparent`,
          borderLeft: `${arrowWidth / 4}px solid ${arrowColor}`,
          marginTop: -arrowWidth / 4,
          marginLeft: width - arrowWidth,
          zIndex: 0
        }}
      />
    </div>
  )
}
Timeline.defaultProps = {
  circleWidth: 50,
  width: 500,
  arrowColor: 'green',
  marginTop: 20
}
export { Timeline, TimelineSVG }
