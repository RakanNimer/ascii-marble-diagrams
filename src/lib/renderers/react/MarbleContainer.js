import React from 'react'
import { Circle } from './Circle'
import { Rectangle } from './Rectangle'

const MarbleContainer = ({
  token: { type, char },
  i,
  width,
  paddingPercentage,
  children
}) => {
  let marble
  const typeWidth = type === 'GROUP' ? width * 8 : width * 2
  switch (type) {
    case 'EMPTY_FRAME':
      marble = null // <Circle r={width}>-</Circle>
      break
    case 'GROUP':
      marble = (
        <Circle r={width}>({char.map(({ char: value }) => value)})</Circle>
      )
      break
    case 'OPERATOR':
      marble = (
        <Rectangle width={typeWidth}>
          {char.map(({ char: value }) => value)}
        </Rectangle>
      )
      break
    default:
      marble = <Circle r={width}>{char}</Circle>
  }

  return (
    <svg
      width={typeWidth}
      height={typeWidth}
      x={width + typeWidth * i + paddingPercentage * width * i}
      y={width}
      key={`marble_${i}`}
    >
      {marble}
    </svg>
  )
}

MarbleContainer.defaultProps = {
  token: { type: 'EMPTY_FRAME', char: 'default' },
  width: 50,
  paddingPercentage: 0.2
}

export { MarbleContainer }
