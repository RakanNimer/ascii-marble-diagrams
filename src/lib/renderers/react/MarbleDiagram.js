import React from 'react'
import { parseMarbleDiagram, tokenizeMarbleDiagram } from '../../'
import { MarbleContainer } from './MarbleContainer'
import { Timeline } from './Timeline'

const MarbleDiagram = ({ width, height, diagram, render }) => {
  const diagramTokens = tokenizeMarbleDiagram(parseMarbleDiagram(diagram))
  const tokenCount = diagramTokens.length > 5 ? diagramTokens.length : 5
  const circleWidth = width / tokenCount / 4
  // const paddingPercentage = 2
  if (render !== null) {
    return render({
      diagramTokens,
      diagram,
      width,
      height,
      marbleRadius: circleWidth * 2
    })
  }
}
MarbleDiagram.defaultProps = {
  width: 1000,
  height: 500,
  diagram: '--c-b-ed-(ad)-',
  render: null
}
export { MarbleDiagram }
