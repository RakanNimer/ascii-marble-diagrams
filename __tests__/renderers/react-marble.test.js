import { createElement } from 'react'
import Enzyme, { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import cases from 'jest-in-case'

import { MarbleDiagram } from '../../src/lib/renderers/react/MarbleDiagram'
Enzyme.configure({ adapter: new Adapter() })

const diagrams = [
  { name: 'empty', diagram: '-' },
  { name: 'one', diagram: '-a-', render: () => createElement('div') },
  { name: 'two consecutive', diagram: '-cb-' },
  { name: 'with error', diagram: '-b-#-' },
  { name: 'hot', diagram: '-^--b--' },
  { name: 'with sync group', diagram: '-(bcd)-' },
  { name: 'with unexpected character', diagram: '->-' },
  { name: 'with invalid group', diagram: '-bcd)-' },
  { name: 'operator', diagram: '[switchMap]' },
  { name: 'invalid operator', diagram: 'switchMap]' },
  { name: 'operator & expression', diagram: '[filter]{i => i < 3}' }
]

describe('react-marble-renderer', () => {
  test('exports', () => {
    expect(MarbleDiagram).toBeDefined()
  })
  cases(
    'mount',
    opts => {
      const element = mount(
        createElement(
          MarbleDiagram,
          { diagram: opts.diagram, render: opts.render },
          null
        )
      )
      expect(element.props()).toMatchSnapshot()
      expect(element.html()).toMatchSnapshot()
    },
    diagrams
  )
})
