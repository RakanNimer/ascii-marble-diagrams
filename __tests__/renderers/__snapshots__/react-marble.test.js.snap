// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-marble-renderer mount empty 1`] = `
Object {
  "children": null,
  "diagram": "-",
  "height": 500,
  "render": null,
  "width": 1000,
}
`;

exports[`react-marble-renderer mount empty 2`] = `null`;

exports[`react-marble-renderer mount hot 1`] = `
Object {
  "children": null,
  "diagram": "-^--b--",
  "height": 500,
  "render": null,
  "width": 1000,
}
`;

exports[`react-marble-renderer mount hot 2`] = `null`;

exports[`react-marble-renderer mount invalid operator 1`] = `
Object {
  "children": null,
  "diagram": "switchMap]",
  "height": 500,
  "render": null,
  "width": 1000,
}
`;

exports[`react-marble-renderer mount invalid operator 2`] = `null`;

exports[`react-marble-renderer mount one 1`] = `
Object {
  "children": null,
  "diagram": "-a-",
  "height": 500,
  "render": [Function],
  "width": 1000,
}
`;

exports[`react-marble-renderer mount one 2`] = `"<div></div>"`;

exports[`react-marble-renderer mount operator & expression 1`] = `
Object {
  "children": null,
  "diagram": "[filter]{i => i < 3}",
  "height": 500,
  "render": null,
  "width": 1000,
}
`;

exports[`react-marble-renderer mount operator & expression 2`] = `null`;

exports[`react-marble-renderer mount operator 1`] = `
Object {
  "children": null,
  "diagram": "[switchMap]",
  "height": 500,
  "render": null,
  "width": 1000,
}
`;

exports[`react-marble-renderer mount operator 2`] = `null`;

exports[`react-marble-renderer mount two consecutive 1`] = `
Object {
  "children": null,
  "diagram": "-cb-",
  "height": 500,
  "render": null,
  "width": 1000,
}
`;

exports[`react-marble-renderer mount two consecutive 2`] = `null`;

exports[`react-marble-renderer mount with error 1`] = `
Object {
  "children": null,
  "diagram": "-b-#-",
  "height": 500,
  "render": null,
  "width": 1000,
}
`;

exports[`react-marble-renderer mount with error 2`] = `null`;

exports[`react-marble-renderer mount with invalid group 1`] = `
Object {
  "children": null,
  "diagram": "-bcd)-",
  "height": 500,
  "render": null,
  "width": 1000,
}
`;

exports[`react-marble-renderer mount with invalid group 2`] = `null`;

exports[`react-marble-renderer mount with sync group 1`] = `
Object {
  "children": null,
  "diagram": "-(bcd)-",
  "height": 500,
  "render": null,
  "width": 1000,
}
`;

exports[`react-marble-renderer mount with sync group 2`] = `null`;

exports[`react-marble-renderer mount with unexpected character 1`] = `
Object {
  "children": null,
  "diagram": "->-",
  "height": 500,
  "render": null,
  "width": 1000,
}
`;

exports[`react-marble-renderer mount with unexpected character 2`] = `null`;
